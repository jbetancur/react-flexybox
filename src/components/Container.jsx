import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { fluid as fluidMedia } from '../utils/media';

const AgnosticComponent = ({
  wrap,
  component,
  fluid,
  ...rest
}) => {
  const Tag = component;

  return <Tag {...rest} />;
};

const ContainerStyle = styled(AgnosticComponent)`
  margin-right: auto;
  margin-left: auto;
  ${props => props.fluid && 'width: 100%;'};
  ${props => !props.fluid && fluidMedia.xs`
    width: 100%;
  `};
  ${props => !props.fluid && fluidMedia.greaterXS`
    max-width: 34rem;
  `};
  ${props => !props.fluid && fluidMedia.greaterSM`
    max-width: 45rem;
  `};
  ${props => !props.fluid && fluidMedia.greaterMD`
    max-width: 60rem;
  `};
  ${props => !props.fluid && fluidMedia.greaterLG`
    max-width: 72.25rem;
  `};
`;

const Container = ({
  fluid,
  className,
  style,
  component,
  children,
  ...rest,
}) => (
  <ContainerStyle
    fluid={fluid}
    className={className}
    style={style}
    component={component}
    {...rest}
  >
    {children}
  </ContainerStyle>
);

Container.propTypes = {
  /**
   * react/html elements
   */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),

  /**
   * The component to render the `Container` as.
   */
  component: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),

  /**
   * full width or controlled width layout
   */
  fluid: PropTypes.bool,

  /**
   * override the `className` on the root element of Row
   */
  className: PropTypes.string,

  /**
   * overide the `style` on the root element of Row
   */
  style: PropTypes.object,
};

Container.defaultProps = {
  children: [],
  fluid: false,
  className: '',
  style: {},
  component: 'div',
};

Container.displayName = 'Container';

export default Container;
