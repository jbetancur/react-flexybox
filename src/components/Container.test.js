import React from 'react';
import { render } from 'react-testing-library';
import 'jest-styled-components';
import Container from './Container';

describe('Container', () => {
  test('it mounts <Container />', () => {
    const { container } = render(<Container />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it mounts <Container fluid />', () => {
    const { container } = render(
      <Container
        fluid
        className="rick-morty"
        style={{ color: 'purple' }}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it mounts <Container component="span" />', () => {
    const { container } = render(
      <Container
        component="span"
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it overrides the default component as a component', () => {
    const Faux = () => <div />;
    const { container } = render(
      <Container
        component={Faux}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
