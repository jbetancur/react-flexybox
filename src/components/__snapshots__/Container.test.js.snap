// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container it mounts <Container /> 1`] = `
.c0 {
  margin-right: auto;
  margin-left: auto;
  width: 100%;
}

@media (min-width:34em) {
  .c0 {
    max-width: 34rem;
  }
}

@media (min-width:48em) {
  .c0 {
    max-width: 45rem;
  }
}

@media (min-width:62em) {
  .c0 {
    max-width: 60rem;
  }
}

@media (min-width:75em) {
  .c0 {
    max-width: 72.25rem;
  }
}

<div
  class="c0"
/>
`;

exports[`Container it mounts <Container component="span" /> 1`] = `
.c0 {
  margin-right: auto;
  margin-left: auto;
  width: 100%;
}

@media (min-width:34em) {
  .c0 {
    max-width: 34rem;
  }
}

@media (min-width:48em) {
  .c0 {
    max-width: 45rem;
  }
}

@media (min-width:62em) {
  .c0 {
    max-width: 60rem;
  }
}

@media (min-width:75em) {
  .c0 {
    max-width: 72.25rem;
  }
}

<span
  class="c0"
/>
`;

exports[`Container it mounts <Container fluid /> 1`] = `
.c0 {
  margin-right: auto;
  margin-left: auto;
  width: 100%;
}

<div
  class="rick-morty c0"
  style="color: purple;"
/>
`;

exports[`Container it overrides the default component as a component 1`] = `<div />`;
